cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java

```diff
@@ -90,65 +99,661 @@
+
+        kcrao.reconcileRebalance(
+                new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                kr).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to PendingProposal (due to the configured Mock server pending calls)
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.PendingProposal);
+                }))).compose(v -> {
+                    // trigger another reconcile to process the PendingProposal state
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+
+                    return kcrao.reconcileRebalance(
+                            new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                            kr1);
+                }).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to ProposalReady
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.ProposalReady);
+                    context.completeNow();
+                })));
+    }
+
+    @Test
+    public void testNewToPendingProposalToStoppedRebalance(Vertx vertx, VertxTestContext context) throws IOException, URISyntaxException {
+
+        // Setup the rebalance endpoint with the number of pending calls before a response is received.
+        MockCruiseControl.setupCCRebalanceResponse(ccServer, 5);
+
+        KafkaRebalance kr =
+                createKafkaRebalance(CLUSTER_NAMESPACE, CLUSTER_NAME, RESOURCE_NAME, new KafkaRebalanceSpecBuilder().build());
+
+        Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).create(kr);
+
+        when(mockKafkaOps.getAsync(CLUSTER_NAMESPACE, CLUSTER_NAME)).thenReturn(Future.succeededFuture(kafka));
+        mockRebalanceOperator(mockRebalanceOps, CLUSTER_NAMESPACE, RESOURCE_NAME, kubernetesClient);
+
+        kcrao.reconcileRebalance(
+                new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                kr).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to PendingProposal (due to the configured Mock server pending calls)
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.PendingProposal);
+                }))).compose(v -> {
```

Comment:
`setHandler()` returns the `Future` on which it is called, not a new `Future`. Calling `compose()` on that sets the handler again, replacing the handler which you set by calling `setHandlder()`.

Reply:
That's true so what do you mean by that? That the lambda specified via the `setHandler` isn't called because overridden by the handler from `compose`. I would agree but the lambda code inside the `setHandler` is actually executed before the one in the `compose`. A similar pattern is used in the `KafkaConnectorIT` as well.

Reply:
Digging into the `FutureImpl` implementation it's not true. The handler in the `compose` doesn't override the previous one defined via `setHandler`, it's added in a collection if there are other handlers.

https://github.com/eclipse-vertx/vert.x/blob/3.8/src/main/java/io/vertx/core/impl/FutureImpl.java#L73

The same is called by the `compose`

https://github.com/eclipse-vertx/vert.x/blob/3.8/src/main/java/io/vertx/core/Future.java#L356

So actually, in the test, the first `setHandler` call sets a handler and the subsequent calls (compose or other setHandler) just adds the handler without overriding the previous ones.

Reply:
Ah, I never knew this. Thank you for correcting me!

Reply:
Well, I just learned it! :-D